FileUtil: move path code to CommonPaths and replace _IDX enum with functions

ligfx opened the pull request:
The large enum of `_IDX` variables always confused me, and seemed more like a leakage of the underlying caching implementation. I asked around on IRC and people seemed open to the idea of changing things, so here's what I came up with.

- Every use of an `_IDX` value has been replaced with it's own getter (and setter, when applicable) function in CommonPaths.h. All backing definitions are in the newly-created CommonPaths.cpp.
- Internally, the large array indexed by `_IDX` values has been replaced with individual static variables.
- Common/FileUtil.h no longer contains any code related to Dolphin paths, and is strictly generic filesystem operations.
- The RebuildUserDirectories function now unconditionally reevaluates all directories and no longer duplicates logic.
- Many defines have been moved out of CommonPaths.h and replaced by constexpr variables in CommonPaths.cpp.

I see a lot of room for improvement, notably with logic that falls back to system directory versions of files, and with consistently templated filenames à la GetThemeDir, but this is already getting pretty big.

leoetlino commented:
Just as a FYI: the Wii stuff should likely use NandPaths instead of CommonPaths directly for consistency. But since it was already that way before this PR, feel free to ignore.

BhaaLseN commented on Source/Android/jni/ButtonManager.cpp:
```diff
@@ -333,7 +334,7 @@ void Init()
   // Init our controller bindings
   IniFile ini;
-  ini.Load(File::GetUserPath(D_CONFIG_IDX) + std::string("Dolphin.ini"));
+  ini.Load(Paths::GetConfigDir() + std::string("Dolphin.ini"));
```
Might as well kill the `std::string` invocations here; the concat operator will do that anyways...(also in the other places)

ligfx replied:
Not only should that `std::string` cast be removed, it should be using `Paths::GetDolphinConfigFile()`!

BhaaLseN commented on Source/Core/Core/Boot/Boot.cpp:
```diff
@@ -132,13 +132,13 @@ bool CBoot::FindMapFile(std::string* existing_map_file, std::string* writable_ma
   bool found = false;
-  static const std::string maps_directories[] = {File::GetUserPath(D_MAPS_IDX),
+  static const std::string maps_directories[] = {Paths::GetMapsDir(),
                                                  File::GetSysDirectory() + MAPS_DIR DIR_SEP};
```
Should `File::GetSysDirectory` be moved over there too?

BhaaLseN commented on Source/Core/Core/State.cpp:
```diff
@@ -326,15 +327,15 @@ static void CompressAndDumpState(CompressAndDumpState_args save_args)
-    if (File::Exists(File::GetUserPath(D_STATESAVES_IDX) + "lastState.sav.dtm"))
-      File::Delete((File::GetUserPath(D_STATESAVES_IDX) + "lastState.sav.dtm"));
+    if (File::Exists(Paths::GetStateSavesDir() + "lastState.sav"))
+      File::Delete((Paths::GetStateSavesDir() + "lastState.sav"));
```
Can we drop the double-braces here while we're at it?

ligfx replied:
Yep.

BhaaLseN commented on Source/Core/Common/CommonPaths.h:
```diff
@@ -235,4 +235,9 @@ const std::string& GetDumpSSLDir()
   return GetUserPath(D_DUMPSSL_IDX);
 }
+
+const std::string& GetLoadDir()
```
Not sure if "Load" is a good name (which also goes for the IDX tho), it doesn't really make sense to me as a term - especially not since it's only used for `titles.txt` (which _is_ being loaded, but thats probably true for half the other files too)

I don't think I can come up with a better name though; so unless someone else has a good idea you can ignore this comment...

ligfx replied:
🤷‍♂️

BhaaLseN commented on Source/Core/Common/CommonPaths.h:
```diff
@@ -260,4 +260,9 @@ const std::string& GetPipesDir()
   return GetUserPath(D_PIPES_IDX);
 }
+
+const std::string& GetMemoryWatcherDir()
```
Is this not used anywhere yet?

ligfx replied:
I guess not, but it existed before so I'm hesitant to remove it.

BhaaLseN commented on Source/Core/Common/CommonPaths.h:
```diff
@@ -280,4 +280,9 @@ const std::string& GetDolphinConfigFile()
   return GetUserPath(F_DOLPHINCONFIG_IDX);
 }
+
+const std::string& GetGCPadConfigFile()
```
And what about this one? This feels a little strange being unused?

BhaaLseN commented on Source/Core/Common/CommonPaths.h:
```diff
@@ -285,4 +285,9 @@ const std::string& GetGCPadConfigFile()
   return GetUserPath(F_GCPADCONFIG_IDX);
 }
+
+const std::string& GetWiimoteConfigFile()
```
And this one too...

BhaaLseN commented on Source/Core/Common/CommonPaths.cpp:
```diff
@@ -0,0 +1,421 @@
+  NUM_PATH_INDICES
+};
+
+static std::string s_user_paths[NUM_PATH_INDICES];
```
`std::array`?

BhaaLseN commented on Source/Core/Common/CommonPaths.cpp:
```diff
@@ -21,6 +21,69 @@
+
+// Files in the directory returned by GetUserPath(D_MEMORYWATCHER_IDX)
+#define MEMORYWATCHER_LOCATIONS "Locations.txt"
+#define MEMORYWATCHER_SOCKET "MemoryWatcher"
```
What about making those `constexpr` instead of `#define`?

BhaaLseN commented on Source/Core/Common/CommonPaths.h:
```diff
@@ -27,89 +26,26 @@
-#define MEMORYWATCHER_SOCKET "MemoryWatcher"
-
 // Sys files
 #define TOTALDB "totaldb.dsy"
```
Can we make those into `constexpr` too?

ligfx replied:
I looked at it, many of these are used in adjacent string literals, so it'd be a really large change. Probably an improvement for a future time.

BhaaLseN replied:
Sure, good enough for me; leave them as-is

BhaaLseN commented on Source/Core/Common/CommonPaths.cpp:
```diff
@@ -0,0 +1,536 @@
+#endif
+
+// Shared data dirs (Sys and shared User for Linux)
+constexpr auto SYSDATA_DIR =
```
What does `auto` boil down to here? `const char*`? And would it be nicer/a better idea/a stupid idea to use a type (like `std::string`) instead?
(comment ignorable, depending on what the answer is)

goldenercobra replied:
I wrote some example code:

``` c++
#include <iostream>
#include <string>
#include <type_traits>

int main() {
  constexpr auto what_am_i = "What am I";
  std::cout<<"Type is: " << typeid(what_am_i).name() << std::endl;

  return 0;
}
```
Compiled with `clang++ -std=c++14 auto_string_test.cpp -o auto_string_test.o` on a Macbook 2014 Early 14 Air, and run with `./auto_string_test.o | c++filt`, yields `char const*`

However, `constexpr std::strings` are not possible due to them not being literals, so `std::string` would not work. There are two solutions; either have a compile-time c-style string which can be constructed into a `std::string` at runtime, or use `std::experimental::string_view`, which is a c++17 feature. 

So, leaving it `const char*` is the best idea imo.

BhaaLseN replied:
Thanks!

BhaaLseN commented on Source/Core/Common/CommonPaths.cpp:
```diff
@@ -0,0 +1,536 @@
+constexpr auto SYSDATA_DIR =
+#if defined(_WIN32) || defined(LINUX_LOCAL_DEV)
+    "Sys"
+#elif defined __APPLE__
```
Is there any reason for being inconsistent, with `_WIN32` and `LINUX_LOCAL_DEV` in braces but `__APPLE__` etc without?

ligfx replied:
It was like that before, I'll consistentify it!

ligfx commented:
@leoetlino Okay, I moved some things into NANDPaths where it made sense (and got rid of some defines from CommonPaths, which is nice). I also moved it to a NANDPaths namespace since it was starting to get a little crowded.

@BhaaLseN Thanks for taking a look! Yeah, I'm trying to stay away from the inconsistencies/path concatenation for now, since they're everywhere and fixing them would make this even larger. Sorry for the huge number of commits, that's the only way I could stay sane doing this 😆

BhaaLseN commented on Source/Android/jni/MainAndroid.cpp:
```diff
@@ -608,7 +609,7 @@ JNIEXPORT jstring JNICALL Java_org_dolphinemu_dolphinemu_NativeLibrary_GetConfig
   std::string defaultValue = GetJString(env, jDefault);
 
-  ini.Load(File::GetUserPath(D_CONFIG_IDX) + std::string(file));
+  ini.Load(Paths::GetConfigDir() + std::string(file));
```
Sorry, I should've probably pointed those out explicitly; but only did the first one (over in `ButtonManager.cpp`) last time...

BhaaLseN commented on Source/Android/jni/MainAndroid.cpp:
```diff
@@ -624,10 +625,10 @@ JNIEXPORT void JNICALL Java_org_dolphinemu_dolphinemu_NativeLibrary_SetConfig(
 
   ini.GetOrCreateSection(section)->Set(key, value);
-  ini.Save(File::GetUserPath(D_CONFIG_IDX) + std::string(file));
+  ini.Save(Paths::GetConfigDir() + std::string(file));
```
Same here and two lines earlier.

lioncash commented on Source/Core/Common/CommonPaths.cpp:
```diff
@@ -0,0 +1,554 @@
+{
+  if (dir.empty())
+    return;
+  s_user_dir = dir;
```
This (and the below setters) can `std::move` the string here.

shuffle2 commented:
plz2rebase

ligfx commented:
I haven't touched this in forever, and I sort of came around to thinking that this is mostly churn and not much benefit. The large number of path concatenations everywhere I think would be a better place to start cleaning up (maybe functions like `GetConfigFile(WIIMOTE_INI_NAME) -> IniFile` instead of `std::string path = GetUserPath(D_CONFIG_IDX) + WIIMOTE_INI_NAME ".ini"`, for instance?)
